sycl/include/sycl/ext/oneapi/experimental/root_group.hpp

```diff
@@ -78,6 +85,75 @@ template <int Dimensions> class root_group {
 
   bool leader() const { return get_local_id() == 0; };
 
+  template <execution_scope Scope, typename RetTy>
+  using checkScopeTy = std::enable_if_t<(Scope == execution_scope::work_item ||
+                                         Scope == execution_scope::sub_group ||
+                                         Scope == execution_scope::work_group),
+                                        RetTy>;
+
+  template <execution_scope Scope>
+  std::enable_if_t<(Scope == execution_scope::work_item ||
+                    Scope == execution_scope::work_group),
+                   id<Dimensions>>
+  get_id() const {
+    if constexpr (Scope == execution_scope::work_item)
+      return it.get_global_id();
+    else if constexpr (Scope == execution_scope::work_group)
+      return it.get_group().get_group_id();
+  }
+
+  template <execution_scope Scope>
+  std::enable_if_t<Scope == execution_scope::sub_group, id<1>> get_id() const {
+    return get_linear_id<execution_scope::sub_group>();
+  }
+
+  template <execution_scope Scope>
+  checkScopeTy<Scope, size_t> get_linear_id() const {
+    if constexpr (Scope == execution_scope::work_item) {
+      return it.get_global_linear_id();
+    } else if constexpr (Scope == execution_scope::sub_group) {
+      size_t WIId = it.get_global_linear_id();
+      size_t SGSize = it.get_sub_group().get_local_linear_range();
```

Comment:
I think this should be `get_max_local_range()[0]`. @Pennycook , do I remember correctly what happens when "last" sub-group is partially masked out?

If so, we'd need a test for this as well.

Reply:
Agree, will update.

Reply:
I'm not 100% sure, but can you do this without a division?



```c++
it.get_group().get_group_linear_id() * it.get_sub_group().get_group_linear_range() + it.get_sub_group().get_group_linear_id()
```

Reply:
It just occurred to me that this would give a different linearization order, and that the linearization order implied by these functions might not be well-defined yet... Different people might expect them to do different things.

Reply:
I think we need some clarification.  We should talk to some other folks and figure out what the right thing to do here is.